### ISSUE TITLE
Type annotations used above their class's definition fail to resolve 

### ISSUE THREAD

**Issue by aThorp96** [microsoft/pyright#5681]
**Describe the bug**
When referencing a class as a type annotation, if the class is defined later in the file then Pyright reports the class in the type annotation is not defined.

**Code or Screenshots**
Example of the error:
```python
class Foo:
    def to_bar(self) -> Bar:
        return Bar()

class Bar:
    pass
```
![Screenshot 2023-08-10 at 1 01 09 PM](https://github.com/microsoft/pyright/assets/28596783/a13e925e-8bbc-450e-b953-ede77816cf9b)

Compared to if the class order is switched:
```python
class Bar:
    pass

class Foo:
    def to_bar(self) -> Bar:
        return Bar() 
```
![Screenshot 2023-08-10 at 1 02 47 PM](https://github.com/microsoft/pyright/assets/28596783/db186482-34ad-4310-8307-17c187b589b3)

**VS Code extension or command-line**
I am using the CLI version 1.1.321, and I have confirmed this is also the case in the VS Code extension of the same version.

**Comment by erictraut** [comment]
This behavior is intended. If you attempt to run the code in your first example, you'll see that the Python interpreter emits an error "NameError: name 'Bar' is not defined". By default, the Python interpreter evaluates type annotations at the point where they are encountered, so if the symbol hasn't yet been defined, you'll see this error.

There are two ways to work around this:
1. Add the statement `from __future__ import annotations` to the top of your file. This tells the Python interpreter to defer evaluation of type annotations.
2. Surround the type annotation in quotes (`"Bar"`) so the Python interpreter evaluates it as a string. PEP 484 indicates that type checkers should parse the contents of quoted type annotations. This provision was added as a way to support forward declarations.

**Comment by aThorp96** [comment]
I see. Would pyright be able to detect that this was the case? If so, it would be really useful to emit a message regarding this. Even if it just mentioned the PEP

**Comment by erictraut** [comment]
Pyright already detects this. You'll see that if you apply one of the two fixes above, the error will go away.

We don't receive many questions about this (maybe one or two a year), so I don't know if it's worth making the error message more verbose. I guess it's something we could consider if there we see an uptick in questions on this topic. The current error message reflects the language in the Python interpreter's error message, so if you attempt to run the code, the problem is usually quite obvious.

Also, I'll note that the `from __future__ import annotations` behavior is slated to become default in Python 3.13 (due out next year), so this issue will likely go away completely over time.

**Comment by aThorp96** [comment]
That makes sense. Thanks!

**Comment by aThorp96** [comment]
Closing because this is intentional
